File: Python/ModelTax.py
# -*- coding: utf-8 -*-
import csv
from io import open

filers = []
opts = []
filer = {}
#    
#def readAllFile(inputFile):
#    try:
#        with open(inputFile, newline = '') as file:
#            reader = csv.reader(file)
#            for row in reader:
#                if (row != ''):
#                    try:
#                        filers.append(TaxFiler())
#                    except Exception as ex :
#                        print('Error TaxFiler {}'.format(type(ex)))
#                    else:
#                        filers[len(filers) - 1].extractVarsFromLine(row)
#    except OSError:
#        print('Failed to Open ' + inputFile)
#    
#def testCalc():
#    findIt = 'RATES1'
#    rates = findValue(findIt)
#    if (rates == -999999):
#        rates = 1
#    for file in filers:
#        file.testOutputVariable = file.salary * rates

#def writeAllOutput(outFile):
#    with open(outFile, 'a') as f:
#        writer = csv.writer(f)
#        for file in filers:
#            writer.writerow(file)

#calculates tax in very simple manner
def calcTax(salary, itemized, standard, rate, exempt, filingStatus):
    tax = rate * (salary - max(itemized, standard) - (1000 * (1 + filingStatus)))
    return tax
 
#testOutputVariable renamed tax
def filerExtractVarsFromLine(inputList):
    taxfiler = {'unitId': inputList[0], 'itemized': float(inputList[1]), 'salary': float(inputList[2]), 'filingStatus':
        int(inputList[3]), 'weight': float(inputList[4]), 'tax': 0.0}
    return taxfiler

def paramExtractVarsFromLine(inputList):
    paramopts = {'label': inputList[0], 'value': float(inputList[1])}
    return paramopts


def readParams(inOpts):
    try:
        with open(inOpts, 'r', newline ='') as file:
            reader = csv.reader(file)
            for row in reader:
                if (row != ''):
                    try:
                        opts.append(paramExtractVarsFromLine(row))
                    except Exception as ex:
                        print('Error ParamOptions {}'.format(type(ex)))
    except OSError:
        print('Failed to Open ' + inOpts)
    
def findValue(label):
    retValue = 999999
    for opt in opts:
        if opt['label'] == label:
            retValue = opt['value']
    return retValue
   
#in main method in c++ version     
def writeOutput(filers, outFile):
    headers = ['unitId', 'itemized', 'salary', 'filingStatus', 'weight', 'tax']
    with open(outFile, 'w', newline = '') as f:
        writer = csv.DictWriter(f, fieldnames=headers)
        writer.writeheader()
        for filer in filers:
            writer.writerow(filer)
        
if __name__ == '__main__':
    print('Python Based Simple Tax Public Model Example')
    inputFile = "..//data//Demo.csv"
    inOpts = "parameterOptions.csv"
    outFile = "output.csv"
    readParams(inOpts)
    count = 0
    try:
        with open(inputFile, 'r', newline = '') as file:
            reader = csv.reader(file)
            for row in reader:
                count += 1
                if (row != ''):
                    filer = filerExtractVarsFromLine(row)
                standard = findValue("STANDARD")
                if (standard == -999999):
                    standard = 1000
                rate = findValue("RATE")
                if (rate == -999999):
                    rate = 0.1
                exempt = findValue("EXEMPT")
                if (exempt == -999999):
                    exempt = 1000
                tax = calcTax(filer['salary'], filer['itemized'], standard, rate, exempt, filer['filingStatus'])
                filer['tax'] = tax
                filers.append(filer)
                if(count%10000 == 0):
                    print('{} observations processed'.format(count))
    except OSError:
        print('Failed to Open ' + inputFile)
        
    print('{} total observations processed'.format(count))
    writeOutput(filers, outFile)
    print('output file ' + outFile + ' written')
    print('processing complete!')

   

File: CPP/ModelTax.cpp
/*********
Toy model. Done in C++
Main entry point.

Jessica A Kelly
jkelly@urban.org

**********/

#include <iostream>
#include <iomanip>
#include <iostream>
#include <fstream>
#include <string>
#include "TaxFiler.h"
#include "ParamOptions.h"
#include "ModelTax.h"
using namespace std;

// Description: Main function to control program flow.
//  Calls read_input, do_regressions and write_output
//  offload these functions to a class
int main(int argc, char* argv[]){
   cout << "C++-based Simple Tax Public Model Example" << endl;
   // create the model
   // could pass as arguments instead
   model = new CppModel("..//data//Demo.csv", "parameterOptions.csv", "output.csv");
   // make sure we are mindful of looping!
   // we could run (but it has more loops than we want):
   // model->readAllFile();
   // model->readParams();
   // model->testCalc();
   // model->writeAllOutput();

   // more efficient with loops
   // read parameters first since will need to know all of these up-front
   model->readParams();

   // begin looping over records as they are read.
   // but the minute we need to know something across all tax filers, this
   // will break down and we need to switch to the above
   ifstream iFile(model->inputFile.c_str());
   model->oFile.open(model->outFile.c_str());
   // write header (could make function)
   model->oFile << "unitId" <<","
     << "itemized" << ","
     << "salary" << ","
     << "filingStatus" << ","
     << "weight" << ","
     << "tax"
     << endl;
   if (iFile.fail()) {
     // do something, this is an error.
     cout << "Failed to open "<< model->inputFile << endl;
   }
   else {
     string FullLine;
     bool firstLine = true; // can use if we have a header.
     long count = 0;
     while (!iFile.eof()) {
       getline(iFile, FullLine);
       //cout << "read a line: "<<FullLine << endl;
       // could put test for header record here.
       firstLine = false;
       count++;
       if (FullLine != "") {
           model->filer = new TaxFiler();
           model->filer->extractVarsFromLine(FullLine);

       }// end full line is not blank

       // after read, then do some math
       // set up to call calcTax
       double salary = model->filer->getSalary();
       double itemized = model->filer->getItemized();
       // if we don't find these options in the parameter file, just set default
       string findIt ="STANDARD";
       double standard = model->findValue(findIt);
       if (standard == -999999) standard = 1000;
       findIt = "RATE";
       double rate = model->findValue(findIt);
       if (rate == -999999) rate = 0.1;
       findIt = "EXEMPT";
       double exempt = model->findValue(findIt);
       if (exempt == -999999) exempt = 1000;
       int filingStatus = model->filer->getFilingStatus();
       // calculate tax
       double tax = model->filer->calcTax(salary, itemized, standard, rate, exempt, filingStatus);
       model->filer->setTestOutputVariable(tax);
       // then we can output
       // could be more elegant here, but not needed.
       model->oFile << model->filer->getUnitId() <<","
         << model->filer->getItemized() << ","
         << model->filer->getSalary() << ","
         << model->filer->getFilingStatus() << ","
         << model->filer->getWeight() << ","
         << model->filer->getTestOutputVariable()
         << endl;
       // add a counter that prints every 10,000 recs read
       if(count%10000 == 0) {
         cout << count << " observations processed." << endl;
       }
     }// end loop over each line
     cout << count << " total observations processed. "<< endl;
     delete model->filer;
   } // end else read file ok

   iFile.close();
   model->oFile.close();
   cout << "output file " << model->outFile.c_str() << " written." << endl;
   cout << "processing complete!" << endl;
   delete model;

   return 0;
}

// all CppModel class below
CppModel::CppModel(const string& inputFilename, const string& inOptsFilename, const string& outputFile):
inputFile(inputFilename), inOpts(inOptsFilename), outFile(outputFile) {
  // any other initialization
}
// let the compiler do it;
//CppModel::~CppModel() {
// delete filers requires a loop
// delete opts requires a loop
//}

// Description: Reads the input file, all in one go.
void CppModel::readAllFile() {
  ifstream iFile(inputFile.c_str());
  if (iFile.fail()) {
    // do something, this is an error.
    cout << "Failed to open "<< inputFile << endl;
  }
  else {
    string FullLine;
    bool firstLine = true; // can use if we have a header.
    long count = 0;
    while (!iFile.eof()) {
      getline(iFile, FullLine);
      //cout << "read a line: "<<FullLine << endl;
      // could put test for header record here.
      firstLine = false;
      if (FullLine != "") {
          try {
            filers.push_back(new TaxFiler());
          }
          catch (exception& oe) {
            cout << "Error TaxFiler " << oe.what() << " count = " << count << endl;
          }
          catch (...)
          {
            cout << "Unknown Error reading tax filer input " << endl;
          }
          int last = filers.size() - 1;
          filers[last]->extractVarsFromLine(FullLine);
        }// end full line is not blank
    }// end loop over each line
  } // end else read file ok

  iFile.close();
}


// Description: reads parameter options and holds those in an options vector.
void CppModel::readParams() {
  ifstream iFile(inOpts.c_str());
  if (iFile.fail()) {
    // do something, this is an error.
    cout << "Failed to open "<< inOpts << endl;
  }
  else {
    string FullLine;
    bool firstLine = true;
    long count = 0;
    while (!iFile.eof()) {
      getline(iFile, FullLine);
      //cout << "read a line: "<<FullLine << endl;
      firstLine = false;
      if (FullLine != "") {
          try {
            opts.push_back(new ParamOptions());
          }
          catch (exception& oe) {
            cout << "Error ParamOptions " << oe.what() << " count = " << count << endl;
          }
          catch (...)
          {
            cout << "Unknown Error reading parameter options input " << endl;
          }
          int last = opts.size() - 1;
          opts[last]->extractVarsFromLine(FullLine);
      }// end full line is not blank
    }// end loop over each line
  } // end else read file ok

  iFile.close();
}

// Description: Use a something from opts and filers
void CppModel::testCalc() {
  string findIt ="RATES1";
  double rates = findValue(findIt);
  if (rates == -999999) rates = 1;
  for(int i =0; i<filers.size(); i++) {
    double testMod = filers[i]->getSalary() * rates;
    filers[i]->setTestOutputVariable(testMod);
    //cout << "doing test calculation and testMod = " << testMod << endl;

  }
}

// Description: writes output all at once.
void CppModel::writeAllOutput(){
  oFile.open(outFile.c_str());
  for(int i =0; i<filers.size(); i++) {
    oFile << filers[i]->getUnitId() <<"," << filers[i]->getTestOutputVariable() <<endl;
  }
  oFile.close();
}

// Description: finds a value corresponding to a label
//  assuming there are only doubles that are values.
//
// Remarks: returns -999999 if cannot find.
double CppModel::findValue(string& label){
  double retValue = -999999;
  for (int i = 0; i< opts.size(); i++) {
    if(opts[i]->getLabel() == label) {
      retValue = opts[i]->getValue();
    }
  }
  return retValue;
}
